samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb

```diff
@@ -0,0 +1,820 @@
+    "# Compute the time evolution on PASQAL emulator\n",
+    "cost_emutn = []\n",
+    "try:\n",
+    "    for T in 1000 * np.linspace(1, 10, 10):\n",
+    "        seq = Sequence(reg, DigitalAnalogDevice)\n",
+    "        seq.declare_channel(\"ising\", \"rydberg_global\")\n",
+    "        adiabatic_pulse = Pulse(\n",
+    "            InterpolatedWaveform(T, [1e-9, Omega, 1e-9]),\n",
+    "            InterpolatedWaveform(T, [delta_0, 0, delta_f]),\n",
+    "            0,\n",
+    "        )\n",
+    "        seq.add(adiabatic_pulse, \"ising\")\n",
+    "\n",
+    "        from collections import Counter\n",
+    "        histogram_results = submit_job(target, seq)\n",
+    "        count_dict = Counter(histogram_results)\n",
+    "\n",
+    "        cost_emutn.append(get_cost(count_dict, Q) / 3)\n",
+    "\n",
+    "except Exception as e:\n",
+    "    print(e)\n",
+    "session.close()"
+   ],
+   "outputs": [],
+   "execution_count": null
+  },
+  {
+   "cell_type": "code",
+   "metadata": {},
+   "source": [
+    "plot_cost(cost_emutn, \"Pasqal Tensor Network-based emulator\")"
+   ],
+   "outputs": [],
+   "execution_count": null
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": "#### Finally, let's compare the results of the two emulators"
+  },
+  {
+   "cell_type": "code",
+   "metadata": {},
+   "source": [
+    "import matplotlib.pyplot as plt\n",
+    "import numpy as np\n",
+    "\n",
+    "def plot_costs(title, costs):\n",
+    "    plt.figure(figsize=(12, 6))\n",
+    "    for label, cost in costs.items():\n",
+    "        plt.plot(range(1, 11), np.array(cost), \"--o\", label=label)\n",
+    "    plt.xlabel(\"total time evolution (µs)\", fontsize=14)\n",
+    "    plt.ylabel(\"cost\", fontsize=14)\n",
+    "    plt.title(title, fontsize=16)\n",
+    "    plt.legend()\n",
+    "    plt.show()"
+   ],
+   "outputs": [],
+   "execution_count": null
+  },
+  {
+   "cell_type": "code",
+   "metadata": {},
+   "source": [
+    "costs = {\n",
+    "    'Qutip emulator': cost_qutip_emulator,\n",
+    "    'Pasqal emulator': cost_emutn\n",
+    "}\n",
+    "plot_costs(\"Qutip emulator vs Pasqal tensor network-based emulator\", costs)"
+   ],
+   "outputs": [],
+   "execution_count": null
+  }
+ ],
+ "metadata": {
+  "celltoolbar": "Tags",
+  "interpreter": {
+   "hash": "949777d72b0d2535278d3dc13498b2535136f6dfe0678499012e853ee9abcab1"
+  },
+  "kernelspec": {
+   "display_name": "Python 3 (ipykernel)",
+   "language": "python",
+   "name": "python3"
+  },
+  "language_info": {
+   "codemirror_mode": {
+    "name": "ipython",
+    "version": 3
+   },
+   "file_extension": ".py",
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.11.2"
+  },
+  "vscode": {
+   "interpreter": {
+    "hash": "e088768f7ff7b4294439f8ed10f7eed9e3b885124bc20d9d06cc2a37b1883330"
+   }
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 4
+}
```

Comment:
So we don't show how to run it on the QPU ? If you replace DigitalAnalogDevice by AnalogDevice and the values of T by the values I provided above, you should have something that can run on Fresnel. I let you see if that's something of interest.

Reply:
I could be a good idea to write another cell with markdown or commented code to explain how to adapt this notebook for Pasqal QPU usage yes, as you mention above.

Given the price of a QPU hour though, I would not recommend adding "runnable QPU code" in these sample notebooks, to make sure the user doesn't get a bad surprise while trying to learn.

Reply:
Okay it was just a suggestion no worries ;)